synedrion/src/cggmp21/sigma/aff_g_star.rs

```diff
@@ -0,0 +1,338 @@
+                let beta = SecretSigned::random_in_exponent_range(rng, P::LP_BOUND + P::EPS_BOUND);
+                let r = Randomizer::random(rng, public.pk0);
+                let s = Randomizer::random(rng, public.pk1);
+
+                let cap_a = (public.cap_c * &alpha + Ciphertext::new_with_randomizer(public.pk0, &beta, &r)).to_wire();
+                let cap_r = secret_scalar_from_signed::<P>(&alpha).mul_by_generator();
+                let cap_b = Ciphertext::new_with_randomizer(public.pk1, &beta, &s).to_wire();
+
+                let ephemeral = AffGStarProofEphemeral::<P> { alpha, beta, r, s };
+                let commitment = AffGStarProofCommitment { cap_a, cap_r, cap_b };
+
+                (ephemeral, commitment)
+            })
+            .unzip();
+
+        let mut reader = XofHasher::new_with_dst(HASH_TAG)
+            // commitments
+            .chain(&commitments)
+            // public parameters
+            .chain(public.pk0.as_wire())
+            .chain(public.pk1.as_wire())
+            .chain(&public.cap_c.to_wire())
+            .chain(&public.cap_d.to_wire())
+            .chain(&public.cap_y.to_wire())
+            .chain(&public.cap_x)
+            .chain(aux)
+            .finalize_to_reader();
+
+        // Non-interactive challenge
+        let e = BitVec::from_xof_reader(&mut reader, P::SECURITY_PARAMETER);
+
+        let elements = ephemerals
+            .into_iter()
+            .zip(e.bits())
+            .map(|(ephemeral, e_bit)| {
+                let z = if *e_bit {
+                    ephemeral.alpha + secret.x
+                } else {
+                    ephemeral.alpha
+                };
+                let z_prime = if *e_bit {
+                    // DEVIATION FROM THE PAPER.
+                    // See the comment in `AffGStarPublicInputs`.
+                    // Original: $+ y$. Modified: $- y$
+                    ephemeral.beta + secret.y.neg()
+                } else {
+                    ephemeral.beta
+                };
+
+                let exponent = if *e_bit {
+                    PublicSigned::one()
+                } else {
+                    PublicSigned::zero()
+                };
+                let w = secret.rho.to_masked(&ephemeral.r, &exponent);
+                // DEVIATION FROM THE PAPER.
+                // See the comment in `AffGStarPublicInputs`.
+                // Original: $\mu^{e_j}$. Modified: $\mu_{-e_j}$.
+                let lambda = secret.mu.to_masked(&ephemeral.s, &-exponent);
+
+                AffGStarProofElement {
+                    z: z.to_public(),
+                    z_prime: z_prime.to_public(),
+                    w,
+                    lambda,
+                }
+            })
+            .collect::<Vec<_>>();
+
+        Self {
+            e,
+            elements: elements.into(),
+            commitments: commitments.into(),
+        }
+    }
+
```

Comment:
Just to re-state: what is a good way to ensure that this list of components is identical to the one in `prove()`? Would be good to have them share code I think.

Reply:
Filed #182